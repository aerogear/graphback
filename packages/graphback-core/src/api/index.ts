
export * from "./ContextCreator"
export * from "./ContextTypes"
export * from "./GraphbackCRUDGeneratorConfig"
export * from "./GraphbackOperationType"