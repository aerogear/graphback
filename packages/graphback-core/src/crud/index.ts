export * from "./mappingHelpers"
export * from "./GraphbackOperationType"
