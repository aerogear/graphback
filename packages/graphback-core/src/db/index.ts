export * from './defaultNameTransforms';
export * from './modelTableMapper';
export * from './dataMapper';
export * from './getPrimaryKey';
