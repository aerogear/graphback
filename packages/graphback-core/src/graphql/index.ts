
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"
export * from "./directives"