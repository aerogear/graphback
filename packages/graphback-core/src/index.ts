
//Top level API
export * from './crud'
export * from './plugin/GraphbackPlugin'
export * from './plugin/GraphbackPluginEngine'
export * from './plugin/GraphbackGlobalConfig'
export * from './plugin/GraphbackCRUDGeneratorConfig';
export * from './plugin/ModelDefinition'
export * from './plugin/getSelectedFieldsFromResolverInfo';
export * from './plugin/GraphbackCoreMetadata'

export * from './relationships/RelationshipMetadataBuilder';
export * from './relationships/relationshipHelpers';
export * from './annotations/DefaultValueAnnotation';

export * from './utils/printSchemaWithDirectives';
export * from './utils/metadataAnnotations';
export * from './utils/fieldTransformHelpers';
export * from './utils/isTransientField';

export * from './runtime';
export * from './db';

export * from './scalars'
