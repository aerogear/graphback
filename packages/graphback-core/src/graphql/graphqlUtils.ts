import { readFileSync } from 'fs';
import { sync } from 'glob';
import { join } from 'path';
import { GraphbackOperationType, InputModelTypeContext, INTERFACE_TYPE_DEFINITION, OBJECT_TYPE_DEFINITION, OBJECT_TYPE_EXTENSION } from '../api';

export const getFieldName = (typeName: string, action: GraphbackOperationType, plural: string = ''): string => {
  const upperCasedType = typeName.charAt(0).toUpperCase() + typeName.substr(1);

  return `${action}${upperCasedType}${plural}`
}

export const getTableName = (typeName: string): string => {
  return typeName.toLowerCase()
}

export const filterObjectTypes = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_DEFINITION);

export const filterInterfaceTypes = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === INTERFACE_TYPE_DEFINITION);

export const filterObjectExtensions = (types: InputModelTypeContext[]) => types.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_EXTENSION);


