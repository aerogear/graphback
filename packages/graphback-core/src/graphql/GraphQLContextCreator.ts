import { parse, visit } from 'graphql';
import { InputModelTypeContext } from '../api/ContextTypes'
import { GraphbackCRUDGeneratorConfig } from "../api/GraphbackCRUDGeneratorConfig";
import { applyGeneratorDirectives } from './directives'
import { filterInterfaceTypes, filterObjectExtensions, filterObjectTypes } from './graphqlUtils';

import { InputContextCreator } from '../api';
import { inputTypeVisitor } from './InputTypeVisitor';



/**
 * Create input context to be reused for:
 * - schema
 * - various generators
 * - database creation
 */
export const graphQLInputContext: InputContextCreator = {

  createModelContext(schemaText: string, defaultConfig: GraphbackCRUDGeneratorConfig): InputModelTypeContext[] {
    const schema = applyGeneratorDirectives(schemaText)

    const astNode = parse(schema)
    const schemaDef = visit(astNode, { leave: inputTypeVisitor });
    const context = applyDefaultConfig(schemaDef, defaultConfig)
    const interfaces = filterInterfaceTypes(context)
    const finalTypes = applyExtensionsToOriginalTypes(context)

    return [...finalTypes, ...interfaces]

  }
}

// tslint:disable-next-line: no-any
function applyExtensionsToOriginalTypes(context: any) {
  const extendNodes = filterObjectExtensions(context)
  const filteredContext = filterObjectTypes(context)

  return filteredContext.map((t: InputModelTypeContext) => {
    const extendNode = extendNodes.find((node: InputModelTypeContext) => node.name === t.name);
    if (extendNode) {
      return {
        ...t,
        fields: [...t.fields, ...extendNode.fields],
        config: { ...t.config, ...extendNode.config },
        interfaces: [...t.interfaces, ...extendNode.interfaces]
      };
    }
    else {
      return t;
    }
  });
}

// tslint:disable-next-line: no-any
function applyDefaultConfig(schemaDef: any, defaultConfig: GraphbackCRUDGeneratorConfig) {
  return schemaDef.definitions.map((t: InputModelTypeContext) => {
    return {
      ...t,
      config: { ...defaultConfig, ...t.config }
    };
  });
}

