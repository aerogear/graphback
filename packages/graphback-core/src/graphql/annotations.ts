import { FieldDefinitionNode, TypeDefinitionNode } from 'graphql';
import { parseAnnotations } from 'graphql-annotations';
import { InputModelFieldAnnotations } from '../api';

export const parseFieldAnnotations = (node: FieldDefinitionNode): InputModelFieldAnnotations => {
    const fieldAnnotations: InputModelFieldAnnotations = {};

    const dbAnnotations: any = node.description ? parseAnnotations('db', String(node.description)) : undefined;

    if (!dbAnnotations) {
        return fieldAnnotations;
    }

    if (dbAnnotations.oneToOne) {
        fieldAnnotations.OneToOne = {
            field: dbAnnotations.oneToOne
        }
    }
    if (dbAnnotations.oneToMany) {
        fieldAnnotations.OneToMany = {
            field: dbAnnotations.oneToMany
        }
    }
    if (dbAnnotations.manyToOne) {
        fieldAnnotations.ManyToOne = {
            field: dbAnnotations.manyToOne
        }
    }
    if (dbAnnotations.manyToMany) {
        fieldAnnotations.ManyToMany = {
            field: dbAnnotations.manyToMany
        }
    }

    return fieldAnnotations;
}


export const parseTypeAnnotations = (node: TypeDefinitionNode) => {

    return node.description ? parseAnnotations('crud', String(node.description)) : undefined;
}
