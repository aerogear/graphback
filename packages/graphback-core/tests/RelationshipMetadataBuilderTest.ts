// eslint-disable-next-line @typescript-eslint/tslint/config
import { buildSchema, GraphQLSchema } from 'graphql';
import { getUserTypesFromSchema } from '@graphql-toolkit/common'
import { RelationshipMetadataBuilder } from '../src';

const setup = (schemaText: string): { builder: RelationshipMetadataBuilder, schema: GraphQLSchema } => {
    const schema = buildSchema(schemaText);
    const modelTypes = getUserTypesFromSchema(schema);
    const builder = new RelationshipMetadataBuilder(modelTypes);

    return { builder, schema };
}

test('should have no relationship metadata', () => {
    const { builder } = setup(`
    type User {
        id: ID!
        name: String
    }`)

    builder.build();

    const userRelationships = builder.getModelRelationships('User');

    expect(userRelationships.length).toEqual(0);
});

test('should create many-to-one relationship metadata from one-to-many field', () => {
    const { builder } = setup(`
    """
    @model
    """
    type User {
        id: ID!
        name: String
        """
        some extra description
        @oneToMany(field: 'sender', key: 'sender_id')
        """
        sentMessages: [Message!]!
    }

    """@model"""
    type Message {
        id: ID!
        description: String
    }
    `)

    builder.build();

    const userRelationships = builder.getModelRelationships('User');
    expect(userRelationships.length).toEqual(1);
    expect(userRelationships[0].kind).toEqual('oneToMany');
    expect(userRelationships[0].ownerField.name).toEqual('sentMessages');
    expect(userRelationships[0].relationFieldName).toEqual('sender');
    expect(userRelationships[0].relationForeignKey).toEqual('sender_id');
    expect(userRelationships[0].relationType.name).toEqual('Message');

    const messageRelationships = builder.getModelRelationships('Message');
    expect(messageRelationships.length).toEqual(1);
    expect(messageRelationships[0].kind).toEqual('manyToOne');
    expect(messageRelationships[0].ownerField.name).toEqual('sender');
    expect(messageRelationships[0].relationFieldName).toEqual('sentMessages');
    expect(userRelationships[0].relationForeignKey).toEqual('sender_id');
    expect(messageRelationships[0].relationType.name).toEqual('User');
});

test('should build one-to-one relationship metadata from one-to-one field', () => {
    const { builder } = setup(`
    """
    @model
    """
    type User {
        id: ID!
        name: String
    }

    """@model"""
    type Address {
        id: ID!
        description: String
        """
        @oneToOne
        """
        resident: User!
    }
    `)

    builder.build();

    const addressRelationships = builder.getModelRelationships('Address');
    expect(addressRelationships.length).toEqual(1);
    expect(addressRelationships[0].kind).toEqual('oneToOne');
    expect(addressRelationships[0].ownerField.name).toEqual('resident');
    expect(addressRelationships[0].relationForeignKey).toEqual('residentId');
    expect(addressRelationships[0].relationType.name).toEqual('User');
    expect(addressRelationships[0].relationFieldName).toBeUndefined();

    expect(JSON.stringify(builder.getRelationships(), undefined, 1)).toMatchSnapshot();
});

test('should build one-to-many and many-to-one relationships from both fields', () => {
    const { builder } = setup(`
    """
    @model
    """
    type User {
        id: ID!
        name: String
        """
        some extra description
        @oneToMany(field: 'sender')
        """
        sentMessages: [Message!]!
    }

    """@model"""
    type Message {
        id: ID!
        description: String
        """
        @manyToOne(field: 'sentMessages')
        """
        sender: User
    }
    `)

    builder.build();

    const userRelationships = builder.getModelRelationships('User');
    expect(userRelationships.length).toEqual(1);
    expect(userRelationships[0].kind).toEqual('oneToMany');
    expect(userRelationships[0].ownerField.name).toEqual('sentMessages')
    expect(userRelationships[0].relationFieldName).toEqual('sender');
    expect(userRelationships[0].relationForeignKey).toEqual('senderId')
    expect(userRelationships[0].relationType.name).toEqual('Message');

    const messageRelationships = builder.getModelRelationships('Message');
    expect(messageRelationships.length).toEqual(1);
    expect(messageRelationships[0].kind).toEqual('manyToOne');
    expect(messageRelationships[0].ownerField.name).toEqual('sender');
    expect(messageRelationships[0].relationFieldName).toEqual('sentMessages');
    expect(userRelationships[0].relationForeignKey).toEqual('senderId')
    expect(messageRelationships[0].relationType.name).toEqual('User');

    expect(JSON.stringify(builder.getRelationships(), undefined, 1)).toMatchSnapshot();
});

test('should throw error when relationship field is missing @model annotation', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToMany(field: 'user')
        """
        messages: [Message]
    }

    type Message {
        id: ID!
        text: String
    }`);

    expect(() => builder.build()).toThrow();
})

test('should throw error when relationship field is not an object type', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToMany(field: 'user')
        """
        messages: [String]
    }`);

    expect(() => builder.build()).toThrow();
});

test('should throw error when relationship field base type is not a model type', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToMany(field: 'user')
        """
        messages: [Message]
    }

    type Message {
        id: ID!
        text: String
    }`);

    expect(() => builder.build()).toThrow();
});

test('should throw error when relation field maps to wrong type', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToMany(field: 'user')
        """
        messages: [Message]
    }

    """@model"""
    type Test {
        id: ID!
    }

    """@model"""
    type Message {
        id: ID!
        text: String
        user: Test
    }`);

    expect(() => builder.build()).toThrow();
});

test('should throw error when relation field is no a single object', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToMany(field: 'user')
        """
        messages: [Message]
    }

    """@model"""
    type Message {
        id: ID!
        text: String
        user: [User]
    }`);

    expect(() => builder.build()).toThrow();
});

test('should throw error when relationship key annotations are not the same', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToMany(field: 'user', key: 'user_id')
        """
        messages: [Message]
    }

    """@model"""
    type Message {
        id: ID!
        text: String
        """
        @manyToOne(field: 'messages', key: 'userId')
        """
        user: User
    }`);

    expect(() => builder.build()).toThrow();
});

test('should throw error when relationship field in one-to-one is not a single object', () => {
    const { builder } = setup(`
    """@model"""
    type User {
        id: ID!
        name: String
        """
        @oneToOne
        """
        messages: [Message]
    }

    """@model"""
    type Message {
        id: ID!
        text: String
    }`);

    expect(() => builder.build()).toThrow();
});

