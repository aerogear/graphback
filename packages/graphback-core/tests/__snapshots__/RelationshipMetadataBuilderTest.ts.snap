// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build one-to-many and many-to-one relationships from both fields 1`] = `
"[
 {
  \\"kind\\": \\"oneToMany\\",
  \\"owner\\": \\"User\\",
  \\"ownerField\\": {
   \\"name\\": \\"sentMessages\\",
   \\"description\\": \\"some extra description\\\\n@oneToMany(field: 'sender', key: 'senderId')\\",
   \\"type\\": \\"[Message!]!\\",
   \\"args\\": [],
   \\"isDeprecated\\": false,
   \\"astNode\\": {
    \\"kind\\": \\"FieldDefinition\\",
    \\"description\\": {
     \\"kind\\": \\"StringValue\\",
     \\"value\\": \\"some extra description\\\\n@oneToMany(field: 'sender')\\",
     \\"block\\": true,
     \\"loc\\": {
      \\"start\\": 89,
      \\"end\\": 171
     }
    },
    \\"name\\": {
     \\"kind\\": \\"Name\\",
     \\"value\\": \\"sentMessages\\",
     \\"loc\\": {
      \\"start\\": 180,
      \\"end\\": 192
     }
    },
    \\"arguments\\": [],
    \\"type\\": {
     \\"kind\\": \\"NonNullType\\",
     \\"type\\": {
      \\"kind\\": \\"ListType\\",
      \\"type\\": {
       \\"kind\\": \\"NonNullType\\",
       \\"type\\": {
        \\"kind\\": \\"NamedType\\",
        \\"name\\": {
         \\"kind\\": \\"Name\\",
         \\"value\\": \\"Message\\",
         \\"loc\\": {
          \\"start\\": 195,
          \\"end\\": 202
         }
        },
        \\"loc\\": {
         \\"start\\": 195,
         \\"end\\": 202
        }
       },
       \\"loc\\": {
        \\"start\\": 195,
        \\"end\\": 203
       }
      },
      \\"loc\\": {
       \\"start\\": 194,
       \\"end\\": 204
      }
     },
     \\"loc\\": {
      \\"start\\": 194,
      \\"end\\": 205
     }
    },
    \\"directives\\": [],
    \\"loc\\": {
     \\"start\\": 89,
     \\"end\\": 205
    }
   }
  },
  \\"relationType\\": \\"Message\\",
  \\"relationFieldName\\": \\"sender\\",
  \\"relationForeignKey\\": \\"senderId\\"
 },
 {
  \\"kind\\": \\"manyToOne\\",
  \\"owner\\": \\"Message\\",
  \\"ownerField\\": {
   \\"name\\": \\"sender\\",
   \\"description\\": \\"@manyToOne(field: 'sentMessages', key: 'senderId')\\",
   \\"type\\": \\"User\\",
   \\"args\\": [],
   \\"isDeprecated\\": false,
   \\"astNode\\": {
    \\"kind\\": \\"FieldDefinition\\",
    \\"description\\": {
     \\"kind\\": \\"StringValue\\",
     \\"value\\": \\"@manyToOne(field: 'sentMessages')\\",
     \\"block\\": true,
     \\"loc\\": {
      \\"start\\": 301,
      \\"end\\": 358
     }
    },
    \\"name\\": {
     \\"kind\\": \\"Name\\",
     \\"value\\": \\"sender\\",
     \\"loc\\": {
      \\"start\\": 367,
      \\"end\\": 373
     }
    },
    \\"arguments\\": [],
    \\"type\\": {
     \\"kind\\": \\"NamedType\\",
     \\"name\\": {
      \\"kind\\": \\"Name\\",
      \\"value\\": \\"User\\",
      \\"loc\\": {
       \\"start\\": 375,
       \\"end\\": 379
      }
     },
     \\"loc\\": {
      \\"start\\": 375,
      \\"end\\": 379
     }
    },
    \\"directives\\": [],
    \\"loc\\": {
     \\"start\\": 301,
     \\"end\\": 379
    }
   }
  },
  \\"relationType\\": \\"User\\",
  \\"relationFieldName\\": \\"sentMessages\\",
  \\"relationForeignKey\\": \\"senderId\\"
 }
]"
`;

exports[`should build one-to-one relationship metadata from one-to-one field 1`] = `
"[
 {
  \\"kind\\": \\"oneToOne\\",
  \\"owner\\": \\"Address\\",
  \\"ownerField\\": {
   \\"name\\": \\"resident\\",
   \\"description\\": \\"@oneToOne(key: 'residentId')\\",
   \\"type\\": \\"User!\\",
   \\"args\\": [],
   \\"isDeprecated\\": false,
   \\"astNode\\": {
    \\"kind\\": \\"FieldDefinition\\",
    \\"description\\": {
     \\"kind\\": \\"StringValue\\",
     \\"value\\": \\"@oneToOne\\",
     \\"block\\": true,
     \\"loc\\": {
      \\"start\\": 176,
      \\"end\\": 209
     }
    },
    \\"name\\": {
     \\"kind\\": \\"Name\\",
     \\"value\\": \\"resident\\",
     \\"loc\\": {
      \\"start\\": 218,
      \\"end\\": 226
     }
    },
    \\"arguments\\": [],
    \\"type\\": {
     \\"kind\\": \\"NonNullType\\",
     \\"type\\": {
      \\"kind\\": \\"NamedType\\",
      \\"name\\": {
       \\"kind\\": \\"Name\\",
       \\"value\\": \\"User\\",
       \\"loc\\": {
        \\"start\\": 228,
        \\"end\\": 232
       }
      },
      \\"loc\\": {
       \\"start\\": 228,
       \\"end\\": 232
      }
     },
     \\"loc\\": {
      \\"start\\": 228,
      \\"end\\": 233
     }
    },
    \\"directives\\": [],
    \\"loc\\": {
     \\"start\\": 176,
     \\"end\\": 233
    }
   }
  },
  \\"relationType\\": \\"User\\",
  \\"relationForeignKey\\": \\"residentId\\"
 }
]"
`;
